Google.Api.Gax.Grpc/ForwardingCallInvoker.cs

```diff
@@ -0,0 +1,155 @@
+        private static readonly Func<Metadata> s_emptyMetadataFunc = () => Metadata.Empty;
+        private static readonly Action s_emptyDisposalAction = () => { };
+
+        private CallInvoker _targetInvoker;
+        private string _sourceMethodFullName;
+        private Method<TTargetRequest, TTargetResponse> _targetMethod;
+        private Func<TSourceRequest, TTargetRequest> _requestConverter;
+        private Func<TSourceRequest, TTargetResponse, TSourceResponse> _responseConverter;
+
+        internal ForwardingCallInvoker(
+            CallInvoker originalInvoker, string sourceMethodFullName, Method<TTargetRequest, TTargetResponse> targetMethod,
+            Func<TSourceRequest, TTargetRequest> requestConverter,
+            Func<TSourceRequest, TTargetResponse, TSourceResponse> responseConverter)
+        {
+            _targetInvoker = originalInvoker;
+            _sourceMethodFullName = sourceMethodFullName;
+            _targetMethod = targetMethod;
+            _requestConverter = requestConverter;
+            _responseConverter = responseConverter;
+        }
+
+        // The streaming methods throw RpcException directly, rather than creating a call that in turn fails.
+        // This is because there are multiple ways of consuming a stream, and throwing directly is more
+        // likely to expose the problem immediately. Also, *all* streaming calls will fail with this
+        // CallInvoker, so making the failure a more catastrophic one is probably appropriate.
+        public override AsyncClientStreamingCall<TRequest, TResponse> AsyncClientStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options) =>
+            throw new RpcException(CreateUnimplementedStatus("Streaming calls are unsupported in this CallInvoker"));
+
+        public override AsyncDuplexStreamingCall<TRequest, TResponse> AsyncDuplexStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options) =>
+            throw new RpcException(CreateUnimplementedStatus("Streaming calls are unsupported in this CallInvoker"));
+
+        public override AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request) =>
+            throw new RpcException(CreateUnimplementedStatus("Streaming calls are unsupported in this CallInvoker"));
+
+        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
+        {
+            if (method.FullName != _sourceMethodFullName)
+            {
+                // Rather than throw directly, here we *do* create a call that fails as if it had been
+                // rejected by a server. This is more likely to be handled appropriately by calling code.
```

Comment:
If this fails, it's a coding error, don't we want to just throw inmediatly? It could literally be ArgumentException I think.

Reply:
I'm torn on this one. I'm worried that there might be plenty of error handling code that expects the initial call to go through, and for the failure to be thrown asynchronously.

Reply:
I just see this as a coding error, not a runtime error. It shouldn't be handle, it should be fix.
But again, I'm fine with this as is, it's just a preference.